aya/src/symbols.rs

```diff
@@ -0,0 +1,688 @@
+//! Utilities for resolving symbols from addresses obtained via stack traces
+use std::{
+    cell::RefCell,
+    collections::BTreeMap,
+    fs::File,
+    io::{self, BufRead, BufReader},
+    path::{Path, PathBuf},
+};
+
+use addr2line::{
+    gimli::{EndianRcSlice, RunTimeEndian},
+    Context,
+};
+use lru_cache::LruCache;
+use thiserror::Error;
+
+use crate::util::kernel_symbols;
+
+/// The error typey returned when creating various [`SymbolResolver`]s.
+#[derive(Debug, Error)]
+pub enum CreateResolverError {
+    /// Failed to create a kernel symbol resolver
+    #[error("Error creating KernelSymbolResolver: {error}")]
+    KernelSymbolResolverBuildError {
+        #[source]
+        /// The original io::Error
+        error: io::Error,
+    },
+}
+
+/// Stores all information that was resolved for a specific symbol
+#[derive(Debug, Clone)]
+pub struct SymbolInfo {
+    /// Virtual memory address of a certain function
+    virtual_address: u64,
+    /// Address of a function inside an object. `None` if it was not resolved
+    object_address: Option<u64>,
+    /// The PID of the process in user space
+    /// `None` is the symbol if the address is a kernel function
+    process_id: Option<u32>,
+    /// The resolved function name
+    function_name: Option<String>,
+    /// Path to the object which defines the symbol
+    object_path: Option<PathBuf>,
```

Comment:
those and potentially even the other fields will be shared a lot between symbols. Could be using `Arc`s? But then OTOH all those strings should be relatively short and there might be more atomic refcount operations, which probably negates the benefit a bit.

Reply:
I think it's probably too soon to think about this - ideally we would borrow from the underlying ELF parser.